pilot/pkg/model/context.go

```diff
@@ -849,14 +858,35 @@ func (node *Proxy) SetServiceInstances(serviceDiscovery ServiceDiscovery) {
 	node.ServiceInstances = instances
 }
 
-// SetWorkloadLabels will set the node.Metadata.Labels only when it is nil.
+// SetWorkloadLabels will set the node.Labels.
+// It merges both node meta labels and workload labels and give preference to workload labels.
 func (node *Proxy) SetWorkloadLabels(env *Environment) {
-	// First get the workload labels from node meta
-	if len(node.Metadata.Labels) > 0 {
+	if len(node.Labels) > 0 {
 		return
 	}
-	// Fallback to calling GetProxyWorkloadLabels
-	node.Metadata.Labels = env.GetProxyWorkloadLabels(node)
+	labels := env.GetProxyWorkloadLabels(node)
+	// when IstioMetaLabels unset, calculate the IstioMetaLabels. It is only set once.
+	if node.IstioMetaLabels == nil {
```

Comment:
This is a bit tricky since we are computing it _once per connection_, not once per proxy lifecycle.

For example. Before state:

```json
  "echo-9bfbfd967-ldf8r.default-7": {
    "Labels": {
      "app": "echo",
      "howardjohn-test": "foo",
      "later": "added",
      "pod-template-hash": "9bfbfd967",
      "security.istio.io/tlsMode": "istio",
      "service.istio.io/canonical-name": "echo",
      "service.istio.io/canonical-revision": "latest",
      "topology.istio.io/cluster": "Kubernetes"
    },
    "IstioMetaLabels": {
      "howardjohn-test": "foo"
    },
    "Metadata.Labels": {
      "app": "echo",
      "howardjohn-test": "foo",
      "security.istio.io/tlsMode": "istio",
      "service.istio.io/canonical-name": "echo",
      "service.istio.io/canonical-revision": "latest"
    }
  },
```

then I remove 'service.istio.io/canonical-name' label, and reconnect to istiod:

```json
  "echo-9bfbfd967-ldf8r.default-1": {
    "Labels": {
      "app": "echo",
      "howardjohn-test": "foo",
      "later": "added",
      "pod-template-hash": "9bfbfd967",
      "security.istio.io/tlsMode": "istio",
      "service.istio.io/canonical-name": "echo",
      "service.istio.io/canonical-revision": "latest",
      "topology.istio.io/cluster": "Kubernetes"
    },
    "MetaLabels": {
      "howardjohn-test": "foo",
      "service.istio.io/canonical-name": "echo"
    },
    "MetadataLabels": {
      "app": "echo",
      "howardjohn-test": "foo",
      "security.istio.io/tlsMode": "istio",
      "service.istio.io/canonical-name": "echo",
      "service.istio.io/canonical-revision": "latest"
    }
  },
```

My Labels hasn't actually changed to account for this; in effect we cannot remove any labels that our pod started with.

I am not quite sure a reliable way to know this though...

Reply:
Right, this is the pre step, which only does refactor part. In #40186, it will update labels.

Reply:
This bug is still present.

Start: a pod has a custom metadata.labels
```
 "shell-858c44699f-knnkq.default-23": {
    "Labels": {
      "app": "shell",
      "cluster": "foo",
      "pod-template-hash": "858c44699f",
      "security.istio.io/tlsMode": "istio",
      "service.istio.io/canonical-name": "shell",
      "service.istio.io/canonical-revision": "latest",
      "topology.istio.io/cluster": "Kubernetes"
    },
    "MetaLabels": {
      "cluster": "foo"
    },
    "MetadataLabels": {
      "app": "shell",
      "cluster": "foo",
      "security.istio.io/tlsMode": "istio",
      "service.istio.io/canonical-name": "shell",
      "service.istio.io/canonical-revision": "latest"
    }
  },

```

Then: `k label pod shell-858c44699f-knnkq service.istio.io/canonical-revision-`
```
    "shell-858c44699f-knnkq.default-23": {
      "Labels": {
        "app": "shell",
        "cluster": "foo",
        "pod-template-hash": "858c44699f",
        "security.istio.io/tlsMode": "istio",
        "service.istio.io/canonical-name": "shell",
        "topology.istio.io/cluster": "Kubernetes"
      },
      "MetaLabels": {
        "cluster": "foo"
      },
      "MetadataLabels": {
        "app": "shell",
        "cluster": "foo",
        "security.istio.io/tlsMode": "istio",
        "service.istio.io/canonical-name": "shell",
        "service.istio.io/canonical-revision": "latest"
      }
    },
```

This is correct so far...

Now, xds disconnects and reconnects:
```
      "shell-858c44699f-knnkq.default-2": {
        "Labels": {
          "app": "shell",
          "cluster": "foo",
          "pod-template-hash": "858c44699f",
          "security.istio.io/tlsMode": "istio",
          "service.istio.io/canonical-name": "shell",
          "service.istio.io/canonical-revision": "latest",
          "topology.istio.io/cluster": "Kubernetes"
        },
        "MetaLabels": {
          "cluster": "foo",
          "service.istio.io/canonical-revision": "latest"
        },
        "MetadataLabels": {
          "app": "shell",
          "cluster": "foo",
          "security.istio.io/tlsMode": "istio",
          "service.istio.io/canonical-name": "shell",
          "service.istio.io/canonical-revision": "latest"
        }
      },
```

Now the label is back unexpectedly